File: hub/recursive/tests/test_bridge_parity.js
'use strict';
/**
 * Cross-language parity test for hub/recursive/bridge.js.
 *
 * Drives the shared `hub/recursive/tests/parity.json` fixture through the
 * Node bridge's pure helpers (`providerArgs`, `globalArgs`, `extractSessionDir`,
 * `lastAssistantText`) and asserts they match the fixture's expectations. The
 * Python bridge runs the same fixture through its helpers in
 * `hub/recursive/tests/test_bridge_parity.py` — together they guard the
 * hub/README.md claim that both bridges produce identical observable behaviour.
 *
 * Run: `node --test hub/recursive/tests/test_bridge_parity.js`
 */

const { test } = require('node:test');
const assert = require('node:assert');
const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');

const HERE = __dirname;
const BRIDGE_JS = path.join(HERE, '..', 'bridge.js');
const PARITY_JSON = path.join(HERE, 'parity.json');

// bridge.js guards main() with `require.main === module`, so requiring it here
// does not run the bridge.
const bridge = require(BRIDGE_JS);
const data = JSON.parse(fs.readFileSync(PARITY_JSON, 'utf8'));

const RECURSIVE_VARS = [
  'RECURSIVE_API_KEY', 'RECURSIVE_PROVIDER', 'RECURSIVE_API_BASE',
  'RECURSIVE_MODEL', 'RECURSIVE_WORKSPACE', 'RECURSIVE_MAX_STEPS',
  'RECURSIVE_PERMISSION_MODE', 'RECURSIVE_STATE_DIR', 'AGENT_STREAMING',
];

function withCleanEnv(caseEnv, fn) {
  const saved = {};
  for (const v of RECURSIVE_VARS) {
    saved[v] = process.env[v];
    delete process.env[v];
  }
  try {
    for (const [k, val] of Object.entries(caseEnv)) {
      process.env[k] = val;
    }
    return fn();
  } finally {
    for (const v of RECURSIVE_VARS) {
      if (saved[v] === undefined) delete process.env[v];
      else process.env[v] = saved[v];
    }
  }
}

for (const c of data.arg_cases) {
  test(`arg_building: ${c.name}`, () => {
    withCleanEnv(c.env, () => {
      assert.deepStrictEqual(bridge.providerArgs(), c.expect_provider_args);
      assert.deepStrictEqual(bridge.globalArgs(), c.expect_global_args);
    });
  });
}

for (let i = 0; i < data.extract_session_dir_cases.length; i++) {
  const c = data.extract_session_dir_cases[i];
  test(`extract_session_dir: ${i}`, () => {
    assert.strictEqual(bridge.extractSessionDir(c.stderr), c.expect);
  });
}

for (let i = 0; i < data.last_assistant_text_cases.length; i++) {
  const c = data.last_assistant_text_cases[i];
  test(`last_assistant_text: ${i}`, () => {
    const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'rec-parity-'));
    if (c.transcript_lines.length) {
      fs.writeFileSync(
        path.join(dir, 'transcript.jsonl'),
        c.transcript_lines.map((l) => JSON.stringify(l)).join('\n') + '\n',
      );
    }
    assert.strictEqual(bridge.lastAssistantText(dir), c.expect);
  });
}

test('last_assistant_text: null session dir', () => {
  assert.strictEqual(bridge.lastAssistantText(null), null);
});

File: hub/recursive/tests/test_bridge_parity.py
"""Cross-language parity test for hub/recursive/bridge.py.

Drives the shared ``hub/recursive/tests/parity.json`` fixture through the
Python bridge's pure helpers (``provider_args``, ``_global_args``,
``extract_session_dir``, ``_last_assistant_text``) and asserts they match the
fixture's expectations. The Node bridge runs the same fixture through its
``providerArgs`` / ``globalArgs`` / ``extractSessionDir`` / ``lastAssistantText``
in ``hub/recursive/tests/test_bridge_parity.js`` — together they guard the
hub/README.md claim that both bridges produce identical observable behaviour.

The full NDJSON event classification (``handleLine`` inside ``main``) is not
covered — it is nested in ``main()`` and not refactored to an importable
helper; that remains future parity work.
"""

from __future__ import annotations

import importlib.util
import json
import os
import sys
from pathlib import Path

import pytest

HERE = Path(__file__).resolve().parent
BRIDGE_PY = HERE.parent / "bridge.py"
PARITY_JSON = HERE / "parity.json"

# Import hub/recursive/bridge.py as a module (it has an __main__ guard, so
# importing it does not run the bridge).
_spec = importlib.util.spec_from_file_location("recursive_bridge_py", BRIDGE_PY)
bridge = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(bridge)

data = json.loads(PARITY_JSON.read_text(encoding="utf-8"))

_RECURSIVE_VARS = [
    "RECURSIVE_API_KEY", "RECURSIVE_PROVIDER", "RECURSIVE_API_BASE",
    "RECURSIVE_MODEL", "RECURSIVE_WORKSPACE", "RECURSIVE_MAX_STEPS",
    "RECURSIVE_PERMISSION_MODE", "RECURSIVE_STATE_DIR", "AGENT_STREAMING",
]


@pytest.fixture
def clean_env(monkeypatch):
    for v in _RECURSIVE_VARS:
        monkeypatch.delenv(v, raising=False)
    yield monkeypatch


@pytest.mark.parametrize("case", data["arg_cases"], ids=[c["name"] for c in data["arg_cases"]])
def test_arg_building(case, clean_env):
    for k, val in case["env"].items():
        clean_env.setenv(k, val)
    assert bridge.provider_args() == case["expect_provider_args"]
    assert bridge._global_args() == case["expect_global_args"]


@pytest.mark.parametrize("case", data["extract_session_dir_cases"],
                         ids=[f"extract-{i}" for i in range(len(data["extract_session_dir_cases"]))])
def test_extract_session_dir(case):
    assert bridge.extract_session_dir(case["stderr"]) == case["expect"]


@pytest.mark.parametrize("case", data["last_assistant_text_cases"],
                         ids=[f"last-{i}" for i in range(len(data["last_assistant_text_cases"]))])
def test_last_assistant_text(case, tmp_path):
    if case["transcript_lines"]:
        (tmp_path / "transcript.jsonl").write_text(
            "\n".join(json.dumps(line) for line in case["transcript_lines"]) + "\n",
            encoding="utf-8",
        )
    assert bridge._last_assistant_text(str(tmp_path)) == case["expect"]


def test_last_assistant_text_none_session_dir():
    assert bridge._last_assistant_text(None) is None
